Make cleaning up pushers depend on the device_id instead of the token_id

sandhose opened the pull request:
Kinda related to https://github.com/matrix-org/synapse/issues/13064#issuecomment-1157644060

This makes it so that we rely on the `device_id` to delete pushers on logout, instead of relying on the `access_token_id`. This ensures we're not removing pushers on token refresh, and prepares for a world without access token IDs (also known as the OIDC)

This actually runs the `set_device_id_for_pushers` background update, which was forgotten in #13831

Note that it currently still delete pushers based on the access_token, so that we don't break the functionality until the background update finishes. I've added comments on all the places where we could remove references to the `access_token_id` once the background update is done.

clokep commented on synapse/handlers/register.py:
```diff
@@ -1013,11 +1013,11 @@ async def _register_email_threepid(
             await self.pusher_pool.add_or_update_pusher(
                 user_id=user_id,
-                access_token=token_id,
+                device_id=device_id,
```
Shouldn't we not provide a `device_id` here? (Otherwise, why is it `Optional`?)

clokep replied:
I think the answer to this is that it is optional, but should almost always be provided. It seems the only time it *isn't* is when a user is registered via the admin API w/ an email.

clokep commented on synapse/push/pusherpool.py:
```diff
@@ -199,19 +203,45 @@ async def remove_pushers_by_app_id_and_pushkey_not_user(
+                await self.remove_pusher(p.app_id, p.pushkey, p.user_name)
+
+    async def remove_http_pushers_by_devices(
+        self, user_id: str, devices: Iterable[str]
```
```suggestion
        self, user_id: str, devices: StrCollection
```

See #14809 (maybe update `remove_http_pushers_by_access_tokens` while we're here?)

clokep commented on synapse/push/pusherpool.py:
```diff
@@ -199,19 +203,45 @@ async def remove_pushers_by_app_id_and_pushkey_not_user(
+        for p in await self.store.get_pushers_by_user_id(user_id):
+            if (
+                p.kind == "http"
+                and p.device_id is not None
+                and p.device_id in device_ids
```
`None` can't be in the input so I think this should work fine?

```suggestion
                and p.device_id in device_ids
```

clokep commented on synapse/storage/databases/main/pusher.py:
```diff
@@ -538,8 +539,8 @@ def set_device_id_for_pushers_txn(txn: LoggingTransaction) -> int:
-                value_names=("device_id",),
-                value_values=[(row["device_id"],) for row in rows],
+                value_names=("device_id", "access_token"),
+                value_values=[(row["device_id"], None) for row in rows],
```
I see the commit says "to make it clear it should now not be used". That might be useful as a comment in the code? 🤷

clokep commented on synapse/push/pusherpool.py:
```diff
@@ -97,7 +97,6 @@ def start(self) -> None:
     async def add_or_update_pusher(
         self,
         user_id: str,
-        access_token: Optional[int],
```
I'm a little confused at the eventual end-goal for `device_id` -- do we think this is now a required property of all pushers? If so, can we update `add_or_update_pusher` to require it (`device_id: str` without a default)?

sandhose replied:
Ideally yes, *but* there is one place where the device_id/access_token_id is not provided: in the admin API, where an admin can register a user without creating a device/access token for the user, that API can create an email pusher on the fly

sandhose commented on synapse/push/pusherpool.py:
```diff
@@ -128,6 +127,22 @@ async def add_or_update_pusher(
         # stream ordering, so it will process pushes from this point onwards.
         last_stream_ordering = self.store.get_room_max_stream_ordering()
 
+        # Before we actually persist the pusher, we check if the user already has one
+        # this app ID and pushkey. If so, we want to keep the access token and device ID
+        # in place, since this could be one device modifying (e.g. enabling/disabling)
+        # another device's pusher.
+        # XXX(quenting): Even though we're not persisting the access_token_id for new
+        # pushers anymore, we still need to copy existing access_token_ids over when
+        # updating a pusher, in case the "set_device_id_for_pushers" background update
+        # hasn't run yet.
```
So this was introduced by [MSC3881](https://github.com/matrix-org/matrix-spec-proposals/pull/3881). With that MSC, you might be updating a pusher that isn't yours, and in that case you want to keep the access_token_id/device_id, and not replace it with the one from the current token.

The only reason we're also carrying the access_token_id here, is that this might happen while the background update hasn't happened yet, and we don't want to break those

clokep replied:
> With that MSC, you might be updating a pusher that isn't yours, and in that case you want to keep the access_token_id/device_id, and not replace it with the one from the current token.

So the `device_id` provided to this function is *always* the device performing the action?

sandhose replied:
It's a bit counterintuitive, but yes. That behaviour was introduced in #13799

sandhose commented:
@clokep I reverted the bugfix on email pushers, since that might need more thinking.

I also changed the background update so that it clears out the access_token column on *all* pushers, instead of just the ones which already had a device_id. I added a bunch of comments there to explain that background update

sandhose commented on synapse/push/pusherpool.py:
```diff
@@ -220,6 +226,26 @@ async def remove_pushers_by_access_token(
+    async def remove_pushers_by_devices(
+        self, user_id: str, devices: StrCollection
+    ) -> None:
+        """Remove the HTTP pushers for a given user corresponding to a set of devices
```
```suggestion
        """Remove the pushers for a given user corresponding to a set of devices
```

clokep commented on synapse/push/pusherpool.py:
```diff
@@ -128,6 +127,22 @@ async def add_or_update_pusher(
+        # pushers anymore, we still need to copy existing access_token_ids over when
+        # updating a pusher, in case the "set_device_id_for_pushers" background update
+        # hasn't run yet.
+        access_token_id = None
+        existing_config = await self._get_pusher_config_for_user_by_app_id_and_pushkey(
+            user_id, app_id, pushkey
+        )
+        if existing_config:
+            device_id = existing_config.device_id
+            access_token_id = existing_config.access_token
```
I think this causes a race with the background update:

1. We fetch an existing pusher config where `device_id` is `None`, but `access_token_id` is set.
2. The background update processes this row, adding the `device_id`.
3. We call `add_pusher` and overwrite the `device_id` back to `None`.

Is that possible?

sandhose replied:
I guess, but it is very unlikely, because unless it happens at the same time as the very last batch of the background update, it will be cleared right after by it
Also, this would only impact old pushers where we don't have the device_id inserted
So, it would have to be at the same time an old enough pusher that doesn't already have a device_id, and a recent enough pusher to be included in the very last batch of background updates

clokep replied:
We talked about this and it *should* be OK since when we add the pusher back (step 3) it actually gets a new ID. So this is an extremely unlikely event to occur.
